osquery/tables/system/windows/certificates.cpp

```diff
@@ -0,0 +1,324 @@
+    }
+  }
+  return join(usages, ",");
+}
+
+void getCertCtxProp(const PCCERT_CONTEXT& certContext,
+                    unsigned long propId,
+                    std::vector<char>& dataBuff) {
+  unsigned long dataBuffLen = 0;
+  auto ret = CertGetCertificateContextProperty(
+      certContext, propId, nullptr, &dataBuffLen);
+  if (ret == 0) {
+    VLOG(1) << "Failed to get certificate property structure " << propId
+            << " with " << GetLastError();
+    return;
+  }
+
+  dataBuff.resize(dataBuffLen, 0);
+  ret = CertGetCertificateContextProperty(
+      certContext, propId, dataBuff.data(), &dataBuffLen);
+
+  if (ret == 0) {
+    VLOG(1) << "Failed to get certificate property struct " << propId
+            << " with " << GetLastError();
+  }
+}
+
+/// Enumerate and process a certificate store
+void enumerateCertStore(const HCERTSTORE& certStore,
+                        const std::string& certStoreName,
+                        std::set<std::string>& processedCerts,
+                        QueryData& results) {
+  auto certContext = CertEnumCertificatesInStore(certStore, nullptr);
+  if (certContext == nullptr) {
+    VLOG(1) << "Failed to enumerate certstore " << certStoreName << " with "
+            << GetLastError();
+    return;
+  }
+
+  while (certContext != nullptr) {
+    // Get the cert fingerprint and ensure we haven't already processed it
+    std::vector<char> certBuff;
+    getCertCtxProp(certContext, CERT_HASH_PROP_ID, certBuff);
+    std::string fingerprint;
+    boost::algorithm::hex(std::string(certBuff.begin(), certBuff.end()),
+                          back_inserter(fingerprint));
+
+    if (processedCerts.find(fingerprint) != processedCerts.end()) {
+      certContext = CertEnumCertificatesInStore(certStore, certContext);
+      continue;
+    }
+    processedCerts.insert(fingerprint);
+
+    Row r;
+    r["sha1"] = fingerprint;
+    certBuff.resize(256);
+    std::fill(certBuff.begin(), certBuff.end(), 0);
+    CertGetNameString(certContext,
+                      CERT_NAME_SIMPLE_DISPLAY_TYPE,
+                      0,
+                      nullptr,
+                      certBuff.data(),
+                      static_cast<unsigned long>(certBuff.size()));
+    r["common_name"] = certBuff.data();
+
+    auto subjSize = CertNameToStr(certContext->dwCertEncodingType,
+                                  &(certContext->pCertInfo->Subject),
+                                  CERT_SIMPLE_NAME_STR,
+                                  nullptr,
+                                  0);
+    certBuff.resize(subjSize);
+    std::fill(certBuff.begin(), certBuff.end(), 0);
+    subjSize = CertNameToStr(certContext->dwCertEncodingType,
+                             &(certContext->pCertInfo->Subject),
+                             CERT_SIMPLE_NAME_STR,
+                             certBuff.data(),
+                             subjSize);
+    r["subject"] = subjSize == 0 ? "" : std::string(certBuff.data());
```

Comment:
In some cases your last argument performs a `std::string` initialization then calls the map's `operator=`, and in others you just pass the assignment constructor a `char*`, as with `certStoreName`. In the case where you create an `rvalue` parameter (in this case) you may be wasting some cycles. I'm not sure if this will be optimized out, or if it will know to call the move assignment for the map's `std::string`.